src/bsp/platform/esp32/xi_bsp_io_fs_esp32.c

```diff
@@ -0,0 +1,307 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <sys/stat.h>
+#include <errno.h>
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <xi_bsp_debug.h>
+#include <xi_bsp_io_fs.h>
+#include <xi_bsp_fwu.h>
+#include <xi_bsp_mem.h>
+
+#include "esp_vfs.h"
+#include "esp_vfs_fat.h"
+#include "esp_ota_ops.h"
+
+#define XI_ESP32_MAX_FILENAME_LEN 64
+#define ESP32_FS_BASE_PATH "/spiflash"
+
+#define XI_BSP_IO_FS_CHECK_CND( cnd, e, s )                                              \
+    if ( ( cnd ) )                                                                       \
+    {                                                                                    \
+        ( s ) = ( e );                                                                   \
+        perror( "Xi FS BSP errno" );                                                     \
+        goto err_handling;                                                               \
+    }
+
+/* These functions are implemented in the application layer to report status */
+extern void esp32_xibsp_notify_update_started( const char* filename, size_t file_size );
+extern void esp32_xibsp_notify_chunk_written( size_t chunk_size, size_t offset );
```

Comment:
I like this solution: wiring out notification to application code to use the device specific LED from there. But what if somebody doesn't link this xilib version to the demo? These functions will be undefined symbols during link time. Still this is a good solution.
But what about solving the same problem with function pointers? This BSP IO FS for ESP32 may define here function pointer types and actual pointers defaulting them NULL. The application may fill in these function pointers in its main.c. Then this BSP code can check against NULL value, if it isn't NULL, then it calls the function, if NULL the BSP does not emit the notification.
This solution lets user link this libxively version to any other application code even which does not define these functions.

Reply:
❗️ what about this?